airflow/providers/google/cloud/utils/credentials_provider.py

```diff
@@ -248,6 +250,16 @@ def get_credentials_and_project(self) -> Tuple[google.auth.credentials.Credentia
 
             project_id = _get_project_id_from_service_account_email(self.target_principal)
 
+        if isinstance(credentials, compute_engine.Credentials):
+            try:
+                credentials.refresh(_http_client.Request())
+            except RefreshError as msg:
+                """
+                If the Compute Engine metadata service can't be reached in this case the instance has not
+                credentials.
+                """
+                self._log_debug(msg)
```

Comment:
This looks like a best-effort approach.

I we were able to retrieve credentials then good, otherwise also good but credentials returned by method will be empty.
It means that the contract of the get_credentials_and_project method is that sometimes it returns you credentials sometimes not, which is in general not a good idea IMHO. At least we need to outline it in the docstring.
Or should we just fail and raise exception with proper message?, especially if it is 4** error which is client error, not backend.

Reply:
Well, then my take is that we have options:
- implement this change with refresh without try/except -> get_credentials_and_project starts to retrieve credentials and some clients/callers which didn't expect this may break (like breeze, and others)
- implement this refresh in different place, so that we fix an original issue and not introduce breaking change (in some sense), e.g. caller that uses "get_credentials_and_project" should do refresh
- have optional parameter in get_credentials_and_project if we want to call refresh or not

I think returning different result in get_credentials_and_project method depending on response error if we were able to retrieve credentials or not is not good interface for caller.

I am curious to hear other opinions.

Reply:
I think we cannot assume we have Metadata server around. This code should work fine whether we have the metadada server or not. IMHO this method should do what it says and return credentials when they are set.

Is there a way to actually check if the credentials we got at this place are ACTUALLY empty? As I understand - we can get the credentials in many ways before - from file, secret, keyfile dictionary or even using "google default" authentication - which does not need refreshing (and that's why Breeze and other places where no meta-data server is available will continue to work - because they got credentials using one of those).

The credentials retrieved this way are perfectly valid. Raising error is not good.

But there is one case that actually **should** throw an error here:

1) none of the methods above were configured
2) default authentication did not provide any authenticatio
3) refresh did not work

In this case, yeah - we should throw an error, as there is no sense continuing and returnign "empty credentials" in this case does not help anyone (and fail-early is the best approach for that).

So - can we throw an error when none-of the above methods actually returned valid credentials and refresh did not do it either?

Reply:
@potiuk after discussion with @kosteev we have decided to move this refresh method to `_get_credentials_email` in `base_google` hook.